[expo-notifications] Extract Firebase managing responsibility to a composite of the central service

sjchmiela opened the pull request:
# Why

A step towards single-threaded notifications handling

# How

- renamed `FirebaseMessagingService` to `NotificationsService` as it will be now the central dispatch station for all notifications-related actions
- ~introduced a service instance property of new interface `FirebaseMessagingDelegate` which mimics Firebase's `onMessageReceived`, `onDroppedMessages` and `onNewToken` methods, but it introduces a new first argument — a `Context`. We want to be able to access `Context` in methods implementations and also to be able to provide `FirebaseMessagingDelegate` as constructor parameter (as per [discussion](https://swmansion.slack.com/archives/CBGPVEWSV/p1601539802006700?thread_ts=1601495523.018300&cid=CBGPVEWSV)) and since the service is the context and we can't initialize the delegate passing in `this` as an argument before `this` has finished initializing (I didn't think `lateinit` or `lazy` would help here too much)~
- introduced a `lazy` service instance property of new interface `FirebaseMessagingDelegate` which mimics Firebase's `onMessageReceived`, `onDroppedMessages` and `onNewToken` methods
- delegated all the methods to the delegate
- extracted scoped logic in `expoview` in the same manner, noarguments constructor of `ExpoFcmMessagingService`

# Test Plan

I have verified manually that push notifications are still received and that it's the `ExpoFcmMessagingService` which governs it in Expo client.

sjchmiela committed:
[expo-notifications] Rename Firebase service to NotificationsService

sjchmiela committed:
[expo-notifications] Extract Firebase responsibilities to a separate delegate

sjchmiela committed:
[android] Move expoview Firebase responsibilities to delegate override

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.java:
```diff
@@ -0,0 +1,99 @@
+            boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(context, remoteMessage);
+              return;
+            }
+          }
+        }
+        // If it's an older experience or useNextNotificationsApi is set to false, let's use the legacy notifications API
+        dispatchToLegacyNotificationModule(context, remoteMessage);
+      } catch (JSONException e) {
+        e.printStackTrace();
+        EXL.e("expo-notifications", "Couldn't parse the manifest.");
+      }
+    } else {
+      EXL.e("expo-notifications", "No experience found for id " + remoteMessage.getData().get("experienceId"));
+    }
+  }
+
+  private void dispatchToNextNotificationModule(Context context, RemoteMessage remoteMessage) {
+    super.onMessageReceived(context, remoteMessage);
+  }
+
+  private void dispatchToLegacyNotificationModule(Context context, RemoteMessage remoteMessage) {
+    PushNotificationHelper.getInstance().onMessageReceived(context, remoteMessage.getData().get("experienceId"), remoteMessage.getData().get("channelId"), remoteMessage.getData().get("message"), remoteMessage.getData().get("body"), remoteMessage.getData().get("title"), remoteMessage.getData().get("categoryId"));
+  }
+
+  @NotNull
+  @Override
+  protected NotificationRequest createNotificationRequest(@NotNull String identifier, @NotNull NotificationContent content, FirebaseNotificationTrigger notificationTrigger) {
+    ExperienceId experienceId;
+    Map<String, String> data = notificationTrigger.getRemoteMessage().getData();
+    if (!data.containsKey("experienceId")) {
+      experienceId = null;
+    } else {
+      experienceId = ExperienceId.create(data.get("experienceId"));
+    }
+    String experienceIdString = experienceId == null ? null : experienceId.get();
+    return new ScopedNotificationRequest(identifier, content, notificationTrigger, experienceIdString);
+  }
+}
```
Hmm, can't we go with kotlin here as well? 🤔 
I believe we're already mixing in the kotlin language into the `expoview`/`kernel` files 😁

sjchmiela replied:
I wished to make review process easier for you by not refactoring this code to Kotlin in the same PR as this would introduce more non-essential changes (and what I did to create `ExpoFcmMessagingDelegate` was simply to change {file,class}name from `ExpoFcmMessagingService` and add `Context context` parameters). I, for one, would prefer to see two separate commits/PRs: one for copy-paste and another one for migration. Seeing that you have already reviewed this and this is the only concern I will migrate to Kotlin in a separate commit. 🙇

bbarthec reviewed (approved):
I'd like to move towards the kotlin supremacy in our codebase more daringly ⚔️ 
I don't see any reason to create more Java files in our codebase at all 😉

bbarthec reviewed (changes requested):
Misclicked the previous one 😓

sjchmiela committed:
[expo-notifications] Remove context argument from delegate methods, use lazily initialized variable instead

bbarthec commented on packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/delegates/FirebaseMessagingDelegate.kt:
```diff
@@ -57,45 +57,38 @@ open class FirebaseListenerService : FirebaseMessagingService() {
-    content: NotificationContent,
-    notificationTrigger: FirebaseNotificationTrigger
-  ): NotificationRequest {
+  protected open fun createNotificationRequest(identifier: String, content: NotificationContent, notificationTrigger: FirebaseNotificationTrigger): NotificationRequest {
```
```suggestion
  protected open fun createNotificationRequest(
    identifier: String,
    content: NotificationContent,
    notificationTrigger: FirebaseNotificationTrigger
  ): NotificationRequest {
```

bbarthec commented on packages/expo-notifications/android/src/main/java/expo/modules/notifications/service/interfaces/FirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,17 @@
+import com.google.firebase.messaging.RemoteMessage
+import expo.modules.notifications.service.NotificationsService
+
+interface FirebaseMessagingDelegate {
```
Do you mind adding some comment for this interface? 🤔 
A bund of words that would address my initial confusion: `what's this interface delegate about?`

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,89 @@
+    )
+  }
+
+  override fun createNotificationRequest(identifier: String, content: NotificationContent, notificationTrigger: FirebaseNotificationTrigger): NotificationRequest {
```
```suggestion
  override fun createNotificationRequest(
    identifier: String,
    content: NotificationContent, 
    notificationTrigger: FirebaseNotificationTrigger
  ): NotificationRequest {
```

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,89 @@
+  private fun dispatchToLegacyNotificationModule(remoteMessage: RemoteMessage) {
+    PushNotificationHelper.getInstance().onMessageReceived(
+      context,
+      remoteMessage.data["experienceId"],
+      remoteMessage.data["channelId"],
+      remoteMessage.data["message"],
+      remoteMessage.data["body"],
+      remoteMessage.data["title"],
+      remoteMessage.data["categoryId"]
```
What would you say if we wrap this `RemoteMessage.data` into our custom data class? 🤔 
I saw that this `RemoteMessage.data` is just a [`Map<string, string>`](https://firebase.google.com/docs/reference/android/com/google/firebase/messaging/RemoteMessage#getData()) and we're accessing every here and there some sophisticated keys in this data payload. I image we can abstract this `Map<string, string>` into some generic `DataClass` with some known properties & provide some inline fallback mechanisms for missing values? 🤔

sjchmiela replied:
If you looked closely you'd notice that this method is marked as _legacy_, in `expo-notifications` we already do it this way — check out `JSONNotificationContentBuilder`!

https://github.com/expo/expo/blob/f36a5161b98f879defd331133604c2677a87b253/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/JSONNotificationContentBuilder.java#L37-L61

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,89 @@
+    if (experienceDBObject != null) {
+      try {
+        val manifest = JSONObject(experienceDBObject.manifest)
+        val sdkVersion = ABIVersion.toNumber(manifest.getString("sdkVersion")) / 10000
```
```suggestion
        val sdkVersion = ABIVersion.majorSdkVersion(manifest.getString("sdkVersion"))
```
Don't we have sth like this? 🤔 
I think we should have such mechanism & delegate all `SDK-version`-related things to that `ABIVersion` class or equivalent.

sjchmiela replied:
I agree this is not the canonical way of comparing SDK versions, compare:

https://github.com/expo/expo/blob/2e6fbac4e7839c9f96fcf35900ec096fbc077a01/android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.java#L51-L55

and

https://github.com/expo/expo/blob/08818849f48b23da2b04cfc0f6641aae733ce877/android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java#L529

I have reverted offending commit and left the code intact.

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,89 @@
+          // In SDK38 and 39 we want to let people decide which notifications API to use,
+          // the next or the legacy one.
+          val androidSection = manifest.optJSONObject("android")
+          if (androidSection != null) {
+            val useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false)
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(remoteMessage)
+              return
+            }
+          }
```
```suggestion
          androidSection?.let {
            val useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false)
            if (useNextNotificationsApi) {
              dispatchToNextNotificationModule(remoteMessage)
              return@let // not sure here though 😅
            }
          }
```
more kotlin-styled 😉

bbarthec commented on android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFirebaseMessagingDelegate.kt:
```diff
@@ -0,0 +1,89 @@
+      return
+    }
+    val experienceDBObject = ExponentDB.experienceIdToExperienceSync(remoteMessage.data["experienceId"])
+    if (experienceDBObject != null) {
```
```suggestion
    experienceDBObject?.let {
```
more kotlin-styled 😉

bbarthec reviewed (commented):
I've got one minor suggestion regarding wrapping `RemoteMessage` in some more friendly abstraction? What would you say?

sjchmiela committed:
[expo-notifications] Add description for FirebaseMessagingDelegate interface

sjchmiela committed:
[expo-notifications] Minor changes to make the diff smaller + prettify

sjchmiela commented:
I removed the commit migrating `expoview` classes from Java to Kotlin as according to your suggestions, @bbarthec, making the migration good would require more than just trivial changes and I wouldn't want to do them as part of this non-related pull request. (plus after last implementation-changing commit Git recognized that `ExpoFcmMessagingService` and `ExpoFirebaseMessagingDelegate` are basically the same and changed the diff from `remove + create` to `rename + modify` which is nicer, more history-friendly and easier to review)

bbarthec reviewed (approved):
Superb 😊 LGTM 👍

sjchmiela committed:
[changelog] Add changelog entry
